src/main/native/include/sysid/analysis/FilteringUtils.h

```diff
@@ -0,0 +1,159 @@
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <vector>
+
+#include "sysid/analysis/AnalysisManager.h"
+
+namespace sysid {
+
+/**
+ * Trims quasistatic data so that no point has a voltage of zero or a velocity
+ * less than the motion threshold.
+ *
+ * @tparam S        The size of the raw data array.
+ * @tparam Voltage  The index of the voltage entry in the raw data.
+ * @tparam Velocity The index of the velocity entry in the raw data.
+ *
+ * @param data            A pointer to the vector of raw data.
+ * @param motionThreshold The velocity threshold under which to delete data.
+ */
+template <size_t S, size_t Voltage, size_t Velocity>
+void TrimQuasistaticData(std::vector<std::array<double, S>>* data,
+                         double motionThreshold) {
+  data->erase(std::remove_if(data->begin(), data->end(),
+                             [motionThreshold](const auto& pt) {
+                               return std::abs(pt[Voltage]) <= 0 ||
+                                      std::abs(pt[Velocity]) < motionThreshold;
+                             }),
+              data->end());
+}
+
+/**
+ * Reduces noise in velocity data by applying a median filter.
+ *
+ * @tparam S The size of the raw data array
+ * @tparam Velocity The index of the velocity entry in the raw data.
+ *
+ * @param data the prepared data vector containing acceleration data
+ * @param window the size of the window of the median filter (must be odd)
+ */
+template <size_t S, size_t Velocity>
+std::vector<std::array<double, S>> ApplyMedianFilter(
+    const std::vector<std::array<double, S>>& data, int window) {
+  if (window % 2 == 0) {
+    throw std::runtime_error("Window must be an odd number");
+  }
+
+  size_t step = window / 2;
+
+  std::vector<std::array<double, S>> prepared;
+
+  // Run median filter on non-edges
+  for (int i = step; i < data.size() - step; i++) {
+    std::vector<std::array<double, S>> velocity_window(window);
+    velocity_window.assign(data.begin() + i - step,
+                           data.begin() + i + step + 1);
+    std::sort(velocity_window.begin(), velocity_window.end(),
+              [&](std::array<double, S> x, std::array<double, S> y) {
+                return x[Velocity] < y[Velocity];
+              });
+    prepared.push_back(std::array<double, S>{data[i]});
+    prepared.back()[Velocity] = velocity_window[step][Velocity];
+  }
+
+  return prepared;
+}
+
+/**
+ * Trims the step voltage data to discard all points before the maximum
+ * acceleration.
+ *
+ * @param data A pointer to the step voltage data.
+ * @param stepTestDuration A reference to the step test duration that will be
+ * used.
+ * @param minStepTime A reference to the minimum step time that will be used.
+ */
+inline void TrimStepVoltageData(std::vector<PreparedData>* data,
```

Comment:
Rather than using `inline`, we should move the implementation of this function to a FilteringUtils.cpp.